static/app/views/explore/components/table.tsx

```diff
@@ -49,38 +52,83 @@ export const ALLOWED_CELL_ACTIONS: Actions[] = [
 
 const MINIMUM_COLUMN_WIDTH = COL_WIDTH_MINIMUM;
 
-type Item = {
-  label: React.ReactNode;
-  value: string;
-  width?: number | 'min-content';
-};
+export function useTableStyles(
+  fields: string[],
+  tableRef: React.RefObject<HTMLDivElement>,
+  minimumColumnWidth = MINIMUM_COLUMN_WIDTH
+) {
+  const resizingColumnIndex = useRef<number | null>(null);
+  const columnWidthsRef = useRef<(number | null)[]>(fields.map(() => null));
 
-interface UseTableStylesOptions {
-  items: Item[];
-  minimumColumnWidth?: number;
-}
+  useEffect(() => {
+    columnWidthsRef.current = fields.map(
+      (_, index) => columnWidthsRef.current[index] ?? null
+    );
+  }, [fields]);
+
+  const initialTableStyles = useMemo(
+    () => ({
+      gridTemplateColumns: fields
+        .map(() => `minmax(${minimumColumnWidth}px, auto)`)
+        .join(' '),
+    }),
+    [fields, minimumColumnWidth]
+  );
+
+  const onResizeMouseDown = useCallback(
+    (event: React.MouseEvent<HTMLDivElement>, index: number) => {
+      event.preventDefault();
+
+      // <GridResizer> is expected to be nested 1 level down from <GridHeadCell>
+      const cell = event.currentTarget!.parentElement;
+      if (!cell) {
+        return;
+      }
+
+      resizingColumnIndex.current = index;
 
-export function useTableStyles({
-  items,
-  minimumColumnWidth = MINIMUM_COLUMN_WIDTH,
-}: UseTableStylesOptions) {
-  const tableStyles = useMemo(() => {
-    const columns = new Array(items.length);
-
-    for (let i = 0; i < items.length; i++) {
-      if (typeof items[i].width === 'number') {
-        columns[i] = `${items[i].width}px`;
-      } else {
-        columns[i] = items[i].width ?? `minmax(${minimumColumnWidth}px, auto)`;
+      const startX = event.clientX;
+      const initialWidth = cell.offsetWidth;
+
+      const gridElement = tableRef.current;
+
+      function onMouseMove(e: MouseEvent) {
+        if (resizingColumnIndex.current === null || !gridElement) {
+          return;
+        }
+
+        const newWidth = Math.max(
+          MINIMUM_COLUMN_WIDTH,
+          initialWidth + (e.clientX - startX)
+        );
+
+        columnWidthsRef.current[index] = newWidth;
+
+        // Updating the grid's `gridTemplateColumns` directly
+        gridElement.style.gridTemplateColumns = columnWidthsRef.current
+          .map(width => {
+            return typeof width === 'number'
+              ? `${width}px`
+              : `minmax(${minimumColumnWidth}px, auto)`;
+          })
+          .join(' ');
       }
-    }
 
-    return {
-      gridTemplateColumns: columns.join(' '),
-    };
-  }, [items, minimumColumnWidth]);
+      function onMouseUp() {
```

Comment:
There's a subtle bug here where another element may have `e.stopPropagation()` on its `onMouseUp` handler and the listeners never get removed. But this should be mostly okay as it will be removed on the next mouse up anywhere.

Reply:
Yup, we can iterate on feedback, good catch though i didn't come across it